infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/ParallelDockerImagesBuilder.java

```diff
@@ -0,0 +1,305 @@
+import org.eclipse.che.workspace.infrastructure.docker.exception.SourceNotFoundException;
+import org.eclipse.che.workspace.infrastructure.docker.logs.MachineLoggersFactory;
+import org.eclipse.che.workspace.infrastructure.docker.model.DockerContainerConfig;
+import org.slf4j.Logger;
+
+/**
+ * This class allows to make parallel prepare (build or download) of docker images for workspace
+ * being started.
+ *
+ * @author Max Shaposhnik (mshaposh@redhat.com)
+ */
+public class ParallelDockerImagesBuilder {
+
+  private static final Logger LOG = getLogger(ParallelDockerImagesBuilder.class);
+  private static final String PARALLEL_PULL_PROPERTY_NAME = "che.infra.docker.max_pull_thread";
+
+  private final RuntimeIdentity identity;
+  private final MachineLoggersFactory machineLoggersFactory;
+  private final boolean doForcePullImage;
+  private final UserSpecificDockerRegistryCredentialsProvider dockerCredentials;
+  private final DockerConnector dockerConnector;
+  private final ThreadPoolExecutor executor;
+
+  @Inject
+  public ParallelDockerImagesBuilder(
+      @Assisted RuntimeIdentity identity,
+      @Named("che.docker.always_pull_image") boolean doForcePullImage,
+      @Named(PARALLEL_PULL_PROPERTY_NAME) int parallelPullsNumber,
+      UserSpecificDockerRegistryCredentialsProvider dockerCredentials,
+      DockerConnector dockerConnector,
+      MachineLoggersFactory machineLoggersFactory) {
+    this.identity = identity;
+    this.doForcePullImage = doForcePullImage;
+    this.dockerCredentials = dockerCredentials;
+    this.dockerConnector = dockerConnector;
+    this.machineLoggersFactory = machineLoggersFactory;
+
+    ThreadFactory factory =
+        new ThreadFactoryBuilder()
+            .setNameFormat(getClass().getSimpleName() + "-%d")
+            .setDaemon(true)
+            .build();
+    this.executor = (ThreadPoolExecutor) Executors.newFixedThreadPool(parallelPullsNumber, factory);
+  }
+
+  /**
+   * Schedule parallel preparation of docker images for the set of docker containers.
+   *
+   * @param containers map of machine name and it's container config
+   * @return map of machine names and theirs image names.
+   * @throws InterruptedException if process is interrupted
+   * @throws InfrastructureException if build failed
+   */
+  public Map<String, String> prepareImages(Map<String, DockerContainerConfig> containers)
+      throws InterruptedException, InfrastructureException {
+    if (executor.getActiveCount() + containers.size() > executor.getMaximumPoolSize()) {
+      LOG.warn(
+          String.format(
+              "Maximum parallel images preparing threads reached. Some images will be queued.\n"
+                  + " Workspace machines count is %s. If problem persists, increase %s property value.",
+              containers.size(), PARALLEL_PULL_PROPERTY_NAME));
+    }
+    Map<String, String> imagesMap = new ConcurrentHashMap<>(containers.size());
```

Comment:
Such variable name doesn't show real variable values. I mean you don't know what it supposed to be as keys and what are values. I'd propose you to rename it to something like `machineToContainer`